### PR TITLE
Improve SQL queries when cleaning orders

### LINKED ISSUE #15202: duplicate key update | ER_LOCK_DEADLOCK: Deadlock found when trying to get lock; try restarting transaction
I am getting duplicate key updates when trying to save data from strapi admin but when I try the same query from tableplus mysql client, it is working fine.

I am surprised where is the duplicate data in that query.
 
attaching a screenshot for detail.
```
insert into `components_elements_pricing_plans_components` (`component_id`, `component_type`, `entity_id`, `field`, `order`) values (15, 'elements.pricing-feature', 3, 'features', 1), (16, 'elements.pricing-feature', 3, 'features', 2), (17, 'elements.pricing-feature', 3, 'features', 3), (18, 'elements.pricing-feature', 3, 'features', 4), (20, 'elements.pricing-feature', 3, 'features', 5), (19, 'elements.pricing-feature', 3, 'features', 6), (21, 'elements.pricing-feature', 3, 'features', 7) on duplicate key update `order` = values(`order`) - ER_LOCK_DEADLOCK: Deadlock found when trying to get lock; try restarting transaction
Error: ER_LOCK_DEADLOCK: Deadlock found when trying to get lock; try restarting transaction
    at Query.Sequence._packetToError (/Users/naazimkhan/Workspace/th-cms/node_modules/mysql/lib/protocol/sequences/Sequence.js:47:14)
    at Query.ErrorPacket (/Users/naazimkhan/Workspace/th-cms/node_modules/mysql/lib/protocol/sequences/Query.js:79:18)
    at Protocol._parsePacket (/Users/naazimkhan/Workspace/th-cms/node_modules/mysql/lib/protocol/Protocol.js:291:23)
    at Parser._parsePacket (/Users/naazimkhan/Workspace/th-cms/node_modules/mysql/lib/protocol/Parser.js:433:10)
    at Parser.write (/Users/naazimkhan/Workspace/th-cms/node_modules/mysql/lib/protocol/Parser.js:43:10)
    at Protocol.write (/Users/naazimkhan/Workspace/th-cms/node_modules/mysql/lib/protocol/Protocol.js:38:16)
    at Socket.<anonymous> (/Users/naazimkhan/Workspace/th-cms/node_modules/mysql/lib/Connection.js:88:28)
    at Socket.<anonymous> (/Users/naazimkhan/Workspace/th-cms/node_modules/mysql/lib/Connection.js:526:10)
    at Socket.emit (node:events:513:28)
    at Socket.emit (node:domain:489:12)
    --------------------
    at Protocol._enqueue (/Users/naazimkhan/Workspace/th-cms/node_modules/mysql/lib/protocol/Protocol.js:144:48)
    at Connection.query (/Users/naazimkhan/Workspace/th-cms/node_modules/mysql/lib/Connection.js:198:25)
    at /Users/naazimkhan/Workspace/th-cms/node_modules/knex/lib/dialects/mysql/index.js:132:18
    at new Promise (<anonymous>)
    at Client_MySQL._query (/Users/naazimkhan/Workspace/th-cms/node_modules/knex/lib/dialects/mysql/index.js:126:12)
    at executeQuery (/Users/naazimkhan/Workspace/th-cms/node_modules/knex/lib/execution/internal/query-executioner.js:37:17)
    at Client_MySQL.query (/Users/naazimkhan/Workspace/th-cms/node_modules/knex/lib/client.js:146:12)
    at /Users/naazimkhan/Workspace/th-cms/node_modules/knex/lib/execution/transaction.js:363:24
    at new Promise (<anonymous>)
    at Client_MySQL.trxClient.query (/Users/naazimkhan/Workspace/th-cms/node_modules/knex/lib/execution/transaction.js:358:12)
```

attaching formated sql:-
```sql
INSERT INTO `components_elements_pricing_plans_components` 
	(`component_id`, `component_type`, `entity_id`, `field`, `order`) values
	(15, 'elements.pricing-feature', 3, 'features', 1), 
	(16, 'elements.pricing-feature', 3, 'features', 2), 
	(17, 'elements.pricing-feature', 3, 'features', 3), 
	(18, 'elements.pricing-feature', 3, 'features', 4), 
	(20, 'elements.pricing-feature', 3, 'features', 5), 
	(19, 'elements.pricing-feature', 3, 'features', 6), 
	(21, 'elements.pricing-feature', 3, 'features', 7);
```
here is the UNIQUE KEY of the table (`component_id`,`component_type`,`entity_id`,`field`)

here is the table schema of the table
```sql
CREATE TABLE `components_elements_pricing_plans_components` (
  `id` int unsigned NOT NULL AUTO_INCREMENT,
  `entity_id` int unsigned DEFAULT NULL,
  `component_id` int unsigned DEFAULT NULL,
  `component_type` varchar(255) DEFAULT NULL,
  `field` varchar(255) DEFAULT NULL,
  `order` int unsigned DEFAULT NULL,
  PRIMARY KEY (`id`),
  UNIQUE KEY `components_elements_pricing_plans_unique` (`entity_id`,`component_id`,`field`,`component_type`),
  KEY `components_elements_pricing_plans_field_index` (`field`),
  KEY `components_elements_pricing_plans_component_type_index` (`component_type`),
  KEY `components_elements_pricing_plans_entity_fk` (`entity_id`),
  CONSTRAINT `components_elements_pricing_plans_entity_fk` FOREIGN KEY (`entity_id`) REFERENCES `components_elements_pricing_plans` (`id`) ON DELETE CASCADE
) ENGINE=InnoDB AUTO_INCREMENT=491 DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_0900_ai_ci;
```

image

image

### LINKED ISSUE #15246: GTID consistency issue
### Required System information

- Node.js version: v16.16.0
- NPM version: 8.11.0
- Strapi version: 4.5.4
- Database: MYSQL 5.7
- Operating system: Alpine Linux 3.13

### Describe the bug

After upgrading from v4.1.7 to v4.5.4 saving any change in the content manager throws an internal server error. We also tried v4.5.3, the same issue happens.

Here are the logs:

`[2022-12-22 11:52:40.732] [31merror[39m: DROP TEMPORARY TABLE IF EXISTS tempOrderTableName_1671709960725 - ER_GTID_UNSAFE_CREATE_DROP_TEMPORARY_TABLE_IN_TRANSACTION: Statement violates GTID consistency: CREATE TEMPORARY TABLE and DROP TEMPORARY TABLE can only be executed outside transactional context.  These statements are also not allowed in a function or trigger because functions and triggers are also considered to be multi-statement transactions.`

`Error: ER_GTID_UNSAFE_CREATE_SELECT: Statement violates GTID consistency: CREATE TABLE ... SELECT.
    at Query.Sequence._packetToError (/app/node_modules/mysql/lib/protocol/sequences/Sequence.js:47:14)
    at Query.ErrorPacket (/app/node_modules/mysql/lib/protocol/sequences/Query.js:79:18)
    at Protocol._parsePacket (/app/node_modules/mysql/lib/protocol/Protocol.js:291:23)
    at Parser._parsePacket (/app/node_modules/mysql/lib/protocol/Parser.js:433:10)
    at Parser.write (/app/node_modules/mysql/lib/protocol/Parser.js:43:10)
    at Protocol.write (/app/node_modules/mysql/lib/protocol/Protocol.js:38:16)
    at Socket.<anonymous> (/app/node_modules/mysql/lib/Connection.js:88:28)
    at Socket.<anonymous> (/app/node_modules/mysql/lib/Connection.js:526:10)
    at Socket.emit (node:events:527:28)
    at Socket.emit (node:domain:475:12)`

It seems to be an issue with GTID consistency.

Turning `@@GLOBAL.GTID_MODE` to OFF fixes this issue. However we don't want to change this because it causes problems on our high availability setup.

Edit: issue seems to be related to the `cleanOrderColumnsForInnoDB` function

https://github.com/strapi/strapi/blob/v4.5.4/packages/core/database/lib/entity-manager/regular-relations.js#L272

### Steps to reproduce the behavior

1. Go to content manager in admin
2. Change a field to something new
3. Click save

### Expected behavior

Saving should work normally.

### LINKED ISSUE #15472: Strapi 4.5.4 very slow put request
### Required System information

- Node.js version: 16.19.0
- NPM version: 8.19.3
- Strapi version: 4.5.4 (4.5.5, 4.5.6)
- Database: Mysql 8.0.25
- Operating system: Ubuntu 22.04.1 LTS, Docker

### Describe the bug

Hello everyone!

Strapi version 4.5.3 worked fine.
But after upgrading to version 4.5.4, it takes a very long time to update if the request contains a 'many-to-many' relation.
Previously, the put request lasted ~100-150ms, and now ~15000-20000ms

If in the file @strapi/database/lib/entity-manager/regular-relations.js comment out the execution of the cleanOrderColumnsForInnoDB function, then everything works fast again.

### Steps to reproduce the behavior

1. Create two collection types
2. Make them "type": "relation", "relation": "manyToMany"
3. Update one of these collection types (there should be a lot of records in the database)
4. The more records there are in the database, the slower the update

### Expected behavior

The request is executed quickly, as before

### Screenshots

With the cleanOrderColumnsForInnoDB function:
Screenshot from 2023-01-17 11-40-33

Without the cleanOrderColumnsForInnoDB function:
Screenshot from 2023-01-17 11-42-02
Screenshot from 2023-01-17 11-41-28

### Code snippets

In my shooting schema.json:
``` 
"employees": {
  "type": "relation",
  "relation": "manyToMany",
  "target": "api::employee.employee",
  "inversedBy": "shootings"
},
```

In my employee schema.json:
```
"shootings": {
  "type": "relation",
  "relation": "manyToMany",
  "target": "api::shooting.shooting",
  "private": true
},
```

### Additional context

In my case I have two collection types (shootings and employees), relation between them "manyToMany", number of records in the database 65469
Screenshot from 2023-01-17 12-16-15

### LINKED ISSUE #15539: Potential performance issue with relational reordering
### Required System information

- Node.js version: v18
- Strapi version: 4.5.6
- Database: All
- Operating system: Windows 10

### Describe the bug

https://github.com/strapi/strapi/blob/127c04705751139bb32805e79b6f8619d92b6d34/packages/core/database/lib/entity-manager/index.js#L605-L634

This uses a select to get the most recent order value and adds 1 to it. This could cause a race condition if you try and add a permission order row twice at the same time because it could depend on old max values.

### PR DESCRIPTION
**Pull request by petersg83**
### What does it do?

Change the way we create order values

### Why is it needed?

The current way is no efficient enough

### How to test it?

You can create a `many-to-many` relation between `category` and `country` and fill an entity with 10000 relations and then removing one and see that the request time is not of 3 minutes anymore.
You need to be on MySQL.
```js
  async bootstrap({ strapi }) {
    const country = await strapi.entityService.create('api::country.country', {
      data: {
        name: 'Spain',
      }
    });
    for (let i=0; i < 10000; i++) {
      await strapi.entityService.create('api::category.category', {
        data: {
          name: `cat-${i}`,
          countries: [country.id],
        }
      });
    }
  },
```

### Related issue(s)/PR(s)
Fix #15246
Fix #15202
Fix #15539
Fix #15472

### CODE REVIEW

File: packages/core/database/lib/entity-manager/regular-relations.js [hunk 1]
```diff
@@ -279,96 +279,70 @@ const cleanOrderColumnsForInnoDB = async ({
   const { joinTable } = attribute;
   const { joinColumn, inverseJoinColumn, orderColumnName, inverseOrderColumnName } = joinTable;
 
-  const now = new Date().valueOf();
-  const randomHex = randomBytes(16).toString('hex');
+  const randomSuffix = `${new Date().valueOf()}_${randomBytes(16).toString('hex')}`;
```
Comment (nathan-pichon) [inline packages/core/database/lib/entity-manager/regular-relations.js:282]: Maybe we can use `crypto.randomUUID()` method ?

Looks like available in v14 (even if MDN says 19? 🤔 )
https://nodejs.org/docs/latest-v14.x/api/crypto.html#crypto_crypto_randomuuid_options

Comment (petersg83) [reply packages/core/database/lib/entity-manager/regular-relations.js:282]: Yes maybe it can be used. I would not risk to change it now however, so the PR is merged for the release of tomorrow

**Comment by Cpaczek** [comment]
I was testing performance on this branch and it works great. However, reordering relationships seems like its broken. Probably why the performance is good.

https://i.imgur.com/AfxRjOE.gif

**Comment by Marc-Roig** [comment]
Nevermind, I was testing with mysql. Mariadb seems to have a problem with reordeirng. Nice catch!

File: packages/core/database/lib/entity-manager/regular-relations.js [hunk 2]
```diff
@@ -279,96 +279,68 @@ const cleanOrderColumnsForInnoDB = async ({
...
+        WHERE a.id = b.id
+        AND a.:joinColumnName: = :id`,
+        {
+          joinTableName: joinTable.name,
+          orderColumnName,
+          joinColumnName: joinColumn.name,
+          id,
+        }
+      )
+      .transacting(trx);
```
Comment (jhoward1994) [inline packages/core/database/lib/entity-manager/regular-relations.js:307]: I think some more descriptive aliases would be nice here. I find it hard to follow what a & b are doing

Comment (Boegie19) [reply packages/core/database/lib/entity-manager/regular-relations.js:307]: Having more descriptive aliases would be nice I agree would make it a lot easier to follow

**Comment by Boegie19** [comment]
@Marc-Roig Is there a way how we can set up a test for this for all dbs you support including master-master config for mariadb so that we can ensure that reodering will not break and we catch any breaking of it before publish in the tests?

**Comment by Marc-Roig** [comment]
@Cpaczek @Boegie19 

We are refactoring for maria db to use row_number, as the other dbs.

Maria db supports row_number() (and windows functions) since 10.2. And supports maria db from 10.3 (tried with 10.1 and strapi doesn't even start).


About testing, we do have automated tests for PG, Sqlite, Mysql (v8). But not for mariadb or mysql 5.7 or in any cluster mode.

We normally test it manually in the strapi project , trying to connect to all dbs. This is something we have to improve for sure.

About performance tests, you are completely on point. We need to define processes about load and performance testing for strapi. To know which are the highest response times, and define at which extend do we allow to degrade API performance. This is on our plans and we will be working on this.

**Comment by Boegie19** [comment]
@Marc-Roig Can you ask if you are allowed to open 2 issues 1 for this
 
1 for mariadb or mysql 5.7 or in any cluster mode. to add checks to them with your implementation plan then the community can give feedback on it + if the plan designed that is good enough maybe  a community member could make the pr  

2 can you also open a issue for performance testing  and like add the information and discussions you guys have had until now in to there aswell just so it is in the open and easier for the me and other community members to follow and help with that?

also if possible maybe this kind of issues could be opened more often so that the community can have input and share there experience

**Comment by Boegie19** [comment]
Tested  mariaDB with the update just made 2 hours ago it works good now just a final node are you guys planning to use decimals? if the answer is no then it would mabybe smart to switch to int so that you don't have the risk of the numers getting to high for a float and that creating an issue. since now in all dbs as soon as you set to a `double unsigned`

**Comment by Marc-Roig** [comment]
> @Marc-Roig Can you ask if you are allowed to open 2 issues

> 1 for mariadb or mysql 5.7 or in any cluster mode. to add checks to them with your implementation plan then the community can give feedback on it + if the plan designed that is good enough maybe a community member could make the pr

Yes! We are holding this for the moment, as github actions are taking a looong time to run. And we are facing 1 hour+ queue times. If we add two or more dbs into the github action matrix that could turn even worse.

We want to look for alternatives to fix this first. But for sure we will ask the community (and u personlly 🚀 ) feedback

> 2 can you also open a issue for performance testing and like add the information and discussions you guys have had until now in to there aswell just so it is in the open and easier for the me and other community members to follow and help with that?

We want to make a formal meeting this week with the team, and we will document this in an issue / RFC for anyone to commen tto :) 

> Tested mariaDB with the update just made 2 hours ago it works good now just a final node are you guys planning to use decimals? if the answer is no then it would mabybe smart to switch to int so it takes less data since now in all dbs as soon as you set to a double unsigned

We do use decimals in an intermediate step when reordering.

When we add a new relation to the DB , the order column has decimals, but we reorder the hole table using ROW_NUMBER afterwards. This method simplifies so much the reordering algo.

File: packages/core/database/lib/entity-manager/regular-relations.js (hunk 2 above)
Comment (petersg83) [reply packages/core/database/lib/entity-manager/regular-relations.js:342]: Thank you, I've added a comment bellow with the SQL as a whole so it can be better understood :)

**Comment by Boegie19** [comment]
@Marc-Roig 
> Yes! We are holding this for the moment, as github actions are taking a looong time to run. And we are facing 1 hour+ queue times. If we add two or more dbs into the github action matrix that could turn even worse.

I agree I onestly also think a RFC or issue should then be opend for the 1 hour+ queue time I personaly see 3 solutions
1 create 2 test runners 1 for in bounds and 1 async for the momend this would not fix it but it would allow to write async tests aswell what will shorten the run time

2 see if we can rewrite our old tests to be async

3 there should be tags in the cease we want to run it agenst all dbs as a temp solution so that that will take longer but we do it on purpese. so that it is atleast posible to test on mariadb or mysql 5.7 or in any cluster mod in actions this should be auto enabled on releases so that we can be sure the release works with all databases

still if you where to make a issue for this I would put this in there awell

**Review by Marc-Roig — APPROVED** [review]

**Comment by Marc-Roig** [comment]
@Boegie19 

I will ask the team what's the common format we use for this public discussions :) And create the issue (or whatever we use)
